### PR TITLE
`MXPX_THRESHOLD` specialized tests

### LINKED ISSUE #2258: `MXPX_THRESHOLD` specialized tests
See https://github.com/Consensys/linea-tracer/pull/2251#discussion_r2333054995 and https://github.com/Consensys/linea-tracer/pull/2251#discussion_r2333130011.

### PR DESCRIPTION
**Pull request by lorenzogentile404**
> [!NOTE]
> Adds comprehensive MXPX threshold tests across key opcodes and replaces standalone MCOPY tests with integrated coverage; minor tweak in random MXP test halting path.
> 
> - **MXP tests**:
>   - **New `MxpxThresholdTests`**: Parameterized boundary tests around `MXPX_THRESHOLD` (London/Cancun) for `MSTORE`/`MSTORE8`, `CODECOPY`, and `CALL`, including overflow scenarios and edge cases where `offset + size` approaches the threshold.
>   - **MCOPY coverage**: Integrates specialized MCOPY tests (post-Cancun) into `MxpxThresholdTests`, replacing `forkSpecific/cancun/McopyTests.java`.
> - **Random MXP test**:
>   - Adjusts halting-path opcode selection in `MxpTest` (TYPE_4 handling).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3a2ba8b03f0028f5527c33dd9411d1c3c580edd4. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java [hunk 1]
```diff
@@ -296,6 +303,69 @@ void testCall(TestInfo testInfo) {
...
+  @ParameterizedTest
+  @ValueSource(ints = {16, 17, 18, 19, 20, 21})
+  void testCodeCopy(int size, TestInfo testInfo) {
+    BytecodeCompiler program = BytecodeCompiler.newProgram(chainConfig);
+    program
+        .push(size)
+        .push(0) // offset (arbitrary value)
+        .push(
+            BigInteger.valueOf(13)
+                .subtract(LONDON_MXPX_THRESHOLD)
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java]: You subtract twice ... typo ?

**Review by OlivierBBB — CHANGES_REQUESTED** [review]
Likely typo in `testCodeCopy`

File: arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java [hunk 2]
```diff
@@ -296,6 +303,69 @@ void testCall(TestInfo testInfo) {
...
+  static Stream<Arguments> testMxpxThresholdSource() {
+    final BigInteger MXPX_THRESHOLD = mxpxThreshold();
+    List<Arguments> arguments = new ArrayList<>();
+    List<BigInteger> values =
+        List.of(
+            BigInteger.ZERO,
+            BigInteger.ONE,
+            BigInteger.valueOf(32),
+            MXPX_THRESHOLD,
+            MXPX_THRESHOLD.add(BigInteger.ONE),
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java]: I think we should have test vectors containing
```rust
// overflow with one tiny parameter
the (a,b) with a = 16, 17, 18 and b = 256**4 - 32 + 15

// overflow with two similar values
the (a,b) with a = 2**31 + {-1, 0, +1} and b = 2**31

// overflow with two large values
the (a,b) with a, b among 256**4 + {-1, 0, +1}
```
and explore both cases
```rust
(offset, size) = (a, b)
(offset, size) = (b, a)
```

Comment (lorenzogentile404) [reply arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java]: by `offset` here you mean any offset parameter or just the parameter of CODECOPY called offset (thus not considering `destOffset`)?

Comment (OlivierBBB) [reply arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java]: No, for CODECOPY the only offset that matters for the MXPX computation is the offset in memory (`destOffset` on EVM codes)

**Review by letypequividelespoubelles — CHANGES_REQUESTED** [review]
MCOPY shoyld ne removed from this tests as its already tested in McooyTests

File: arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java [hunk 3]
```diff
@@ -296,6 +304,218 @@ void testCall(TestInfo testInfo) {
...
+    BytecodeCompiler program = BytecodeCompiler.newProgram(chainConfig);
+
+    switch (opCode) {
+        // 1 offset
+      case MSTORE, MSTORE8 -> program
+          .push(0) // value
+          .push(offset1)
+          .op(opCode);
+        // 2 offsets, 1 size
+      case CODECOPY -> program.push(size1).push(offset2).push(offset1).op(opCode);
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java]: This isn't a 2 offset 1 size opcode from the `MXP` point of view: `MCOPY` is the only such opcode.

File: arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java [hunk 4]
```diff
@@ -296,6 +304,218 @@ void testCall(TestInfo testInfo) {
...
+      default -> throw new IllegalArgumentException("Unsupported opCode: " + opCode);
+    }
+
+    BytecodeRunner.of(program.compile()).run(chainConfig, testInfo);
+  }
+
+  static final BigInteger MAX_UINT256 =
+      BigInteger.TWO.pow(256).subtract(BigInteger.ONE); // 2^256 - 1
+  static final BigInteger LONDON_MXPX_THRESHOLD = (BigInteger.valueOf(256).pow(4));
+  static final BigInteger CANCUN_MXPX_THRESHOLD =
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java]: Are these thresholds not part of the `constants` ?

File: arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java [hunk 5]
```diff
@@ -296,6 +304,218 @@ void testCall(TestInfo testInfo) {
...
+      (BigInteger.valueOf(256).pow(4)).subtract(BigInteger.ONE);
+  static final BigInteger SMALL = BigInteger.valueOf(32);
+
+  static final List<OpCode> oneOffsetOpCodes = List.of(OpCode.MSTORE, OpCode.MSTORE8);
+  static final List<OpCode> twoOffsetsOneSizeOpCodes = List.of(OpCode.CODECOPY);
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java]: Same issue: only `MCOPY` is a 2 offsets, 1 size opcode from the MXP point of view.

File: arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java [hunk 6]
```diff
@@ -296,6 +304,218 @@ void testCall(TestInfo testInfo) {
...
+    return arguments.stream();
+  }
+
+  static Stream<Arguments> testCodeCopyOverflowWithTwoSimilarValuesSource() {
+    List<Arguments> arguments = new ArrayList<>();
+    List<BigInteger> aValues =
+        List.of(
+            BigInteger.TWO.pow(31),
+            BigInteger.TWO.pow(31).add(BigInteger.valueOf(1)),
+            BigInteger.TWO.pow(31).add(BigInteger.valueOf(1)));
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java]: You are adding twice rather than I guess adding and subtracting.

File: arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java [hunk 7]
```diff
@@ -296,6 +304,218 @@ void testCall(TestInfo testInfo) {
     toyExecutionEnvironmentV2.run();
   }
 
+  @Tag("nightly")
+  @ParameterizedTest
+  @MethodSource({"testMxpxThresholdSource"})
+  void testMxpxThreshold(
```
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java]: It would be good to add documentation to say what it is we are testing. Something along the lines of

```java
/**
The following tests were written to test a bug fix in the tracer <add link to PR>
The `HUB` and the `MXP` modules used to have different,
incompatible criteria by which they recognized of memory
expansion exception. One issue was that the London `HUB`
compared a sum `offset + size` to some threshold, while the
`MXP` module did the same but with a different comparison
operator (≤ vs <). The issue arose whenever opcodes had
_precisely_ `offset + size = 256**4`.

Another issue (IIRC, to be confirmed) was that the Cancun `HUB`
still used sums `offset + size` as its gauge rather than the individual
offsets and sizes.

As such we are testing with parameters in that neighbourhood.
We distinguish 3 classes of MXP instructions
- constant size opcodes (e.g. `MSTORE`, `MSTORE8` with respective 'sizes" 32 and 1)
- variable size opcodes
  - single offset, single size (e.g. `RETURN`, `CODECOPY`, ...)
  - double offset, single size (only `MCOPY`)
  - double offset, double size (the `CALL`'s)
*/
```

File: arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java
Comment (OlivierBBB) [inline arithmetization/src/test/java/net/consensys/linea/zktracer/module/mxp/MxpTest.java:1]: These tests should be named `MxpxThresholdTests` or so.

**Review by OlivierBBB — COMMENTED** [review]
Left various comments.

**Review by OlivierBBB — APPROVED** [review]
